consensus/hotstuff/verification/combined_verifier_v2.go

```diff
@@ -0,0 +1,142 @@
+	// TODO: move DecodeDoubleSig to merger.Split
+	stakingSig, beaconShare, err := signature.DecodeDoubleSig(sigData)
+	if err != nil {
+		return false, fmt.Errorf("could not split signature: %w", modulesig.ErrInvalidFormat)
+	}
+
+	dkg, err := c.committee.DKG(block.BlockID)
+	if err != nil {
+		return false, fmt.Errorf("could not get dkg: %w", err)
+	}
+
+	// verify each signature against the message
+	// TODO: check if using batch verification is faster (should be yes)
+	stakingValid, err := signer.StakingPubKey.Verify(stakingSig, msg, c.staking)
+	if err != nil {
+		return false, fmt.Errorf("internal error while verifying staking signature: %w", err)
+	}
+	if !stakingValid {
+		return false, fmt.Errorf("invalid staking sig")
+	}
+
+	// there is no beacon share, no need to verify it
+	if beaconShare == nil {
+		return true, nil
+	}
+
+	// if there is beacon share, there must be beacon public key
+	beaconPubKey, err := dkg.KeyShare(signer.NodeID)
+	if err != nil {
+		return false, fmt.Errorf("could not get random beacon key share for %x: %w", signer.NodeID, err)
+	}
+
+	beaconValid, err := beaconPubKey.Verify(beaconShare, msg, c.beacon)
+	if err != nil {
+		return false, fmt.Errorf("internal error while verifying beacon signature: %w", err)
+	}
+
+	if !beaconValid {
+		return false, fmt.Errorf("invalid beacon sig")
+	}
+	return true, nil
+}
+
+// VerifyQC verifies the validity of a combined signature on a quorum certificate.
+func (c *CombinedVerifierV2) VerifyQC(signers flow.IdentityList, sigData []byte, block *model.Block) (bool, error) {
+
+	dkg, err := c.committee.DKG(block.BlockID)
+	if err != nil {
+		return false, fmt.Errorf("could not get dkg: %w", err)
+	}
+
+	// TODO: to be replaced by packer.Unpack method
+	// split the aggregated staking & beacon signatures
+	stakingAggSig, beaconThresSig, err := c.merger.Split(sigData)
+	if err != nil {
+		return false, fmt.Errorf("could not split signature: %w", modulesig.ErrInvalidFormat)
+	}
+
+	msg := MakeVoteMessage(block.View, block.BlockID)
+	// TODO: verify if batch verification is faster
+
+	// verify the beacon signature first
+	beaconValid, err := dkg.GroupKey().Verify(beaconThresSig, msg, c.beacon)
+	if err != nil {
+		return false, fmt.Errorf("internal error while verifying beacon signature: %w", err)
+	}
+	if !beaconValid {
+		return false, nil
+	}
+	// verify the aggregated staking signature next (more costly)
+	// TODO: eventually VerifyMany will be a method of a stateful struct. The struct would
+	// hold the message, all the participants keys, the latest verification aggregated public key,
+	// as well as the latest list of signers (preferably a bit vector, using indices).
+	// VerifyMany would only take the signature and the new list of signers (a bit vector preferably)
+	// as inputs. A new struct needs to be used for each epoch since the list of participants is upadted.
+
+	aggregatedKey, err := c.keysAggregator.aggregatedStakingKey(signers)
```

Comment:
This is using a module that is going to be deleted in V2, the module that should be used to aggregate public keys is module/signature.PublicKeyAggregator. We can make this change in another PR.
```suggestion
       // TODO: update to use module/signature.PublicKeyAggregator
	aggregatedKey, err := c.keysAggregator.aggregatedStakingKey(signers)
```